Visualize cardinality of foreign key columns

R-Palazzo opened the pull request:
Resolve #283

fealho reviewed (changes requested):
If I'm understanding things correctly, `get_cardinality` can be simplified a bit. Besides that it looks pretty good 👍

R-Palazzo commented:
Thanks for the review @fealho! I addressed the comment in [5ed67c3](https://github.com/sdv-dev/SDMetrics/pull/384/commits/5ed67c3b9fde229f639062099fc701dc3717c5c4)

fealho commented on sdmetrics/reports/utils.py:
```diff
@@ -479,6 +479,124 @@ def get_column_pair_plot(real_data, synthetic_data, column_names, metadata):
+    return fig
+
+
+def get_cardinality_plot(real_data, synthetic_data, child_table_name, parent_table_name,
```
Add `Return` to docstring of all three methods.

R-Palazzo replied:
Yes done in [1de3440](https://github.com/sdv-dev/SDMetrics/pull/384/commits/1de3440d8b0a3f80de23c5d3b0ec909796e23242)

fealho reviewed (commented):
Thanks for addressing, the code looks much more concise now! I tested locally and it seems good, I just wanted to ask you why are you doing the `'# parents'` stuff? It doesn't seem like it gets added to the final image.

R-Palazzo commented:
> Thanks for addressing, the code looks much more concise now! I tested locally and it seems good, I just wanted to ask you why are you doing the `'# parents'` stuff? It doesn't seem like it gets added to the final image.

The `'# parents'` column corresponds to the y-axis in the graph, but we plot the density rather than a count.

amontanez24 commented on sdmetrics/reports/utils.py:
```diff
@@ -479,6 +479,133 @@ def get_column_pair_plot(real_data, synthetic_data, column_names, metadata):
+    Returns:
+        plotly.graph_objects._figure.Figure
+    """
+    fig = px.histogram(data,
+                       x='# children',
+                       y='# parents',
+                       color='data',
+                       barmode='group',
+                       color_discrete_sequence=[DATACEBO_DARK, DATACEBO_LIGHT],
+                       pattern_shape='data',
+                       pattern_shape_sequence=['', '/'],
+                       nbins=max(data['# children']) - min(data['# children']) + 1,
+                       histnorm='probability density')
```
minor: can we start all the parameters a line below and not as tabbed over

R-Palazzo replied:
Yes, done in [76fcf86](https://github.com/sdv-dev/SDMetrics/pull/384/commits/76fcf862286df85acac9ee2977216ac5c72f5a30)

amontanez24 commented on sdmetrics/reports/utils.py:
```diff
@@ -479,6 +479,133 @@ def get_column_pair_plot(real_data, synthetic_data, column_names, metadata):
         return make_mixed_column_pair_plot(real_data, synthetic_data)
 
 
+def get_cardinality(parent_table, child_table, parent_primary_key, child_foreign_key):
```
can we make this private for now?

R-Palazzo replied:
Yes I was thinking of this first, but then I saw that we don't do it in this file. Done in [76fcf86](https://github.com/sdv-dev/SDMetrics/pull/384/commits/76fcf862286df85acac9ee2977216ac5c72f5a30)

amontanez24 commented on sdmetrics/reports/utils.py:
```diff
@@ -479,6 +479,133 @@ def get_column_pair_plot(real_data, synthetic_data, column_names, metadata):
+    return cardinalities.sort_values('# children')
+
+
+def generate_cardinality_plot(data, parent_primary_key, child_foreign_key):
```
can we make this private too?

R-Palazzo replied:
Yes, done in [76fcf86](https://github.com/sdv-dev/SDMetrics/pull/384/commits/76fcf862286df85acac9ee2977216ac5c72f5a30)

amontanez24 commented on sdmetrics/reports/utils.py:
```diff
@@ -479,6 +479,133 @@ def get_column_pair_plot(real_data, synthetic_data, column_names, metadata):
+        plotly.graph_objects._figure.Figure
+    """
+    parent_primary_key = None
+    for relation_dict in metadata.get('relationships', []):
```
minor: I don't think we need the dict suffix

R-Palazzo replied:
Yes, done in [76fcf86](https://github.com/sdv-dev/SDMetrics/pull/384/commits/76fcf862286df85acac9ee2977216ac5c72f5a30)

amontanez24 commented on sdmetrics/reports/utils.py:
```diff
@@ -479,6 +479,133 @@ def get_column_pair_plot(real_data, synthetic_data, column_names, metadata):
+            parent_primary_key = relation_dict['parent_primary_key']
+
+    if parent_primary_key is None:
+        raise ValueError(
+            f"Relationship between child table '{child_table_name}' and parent table "
+            f"'{parent_table_name}' for the foreign key '{child_foreign_key}' "
+            'not found in the metadata. Please update the metadata.'
```
minor: can we reword this to say "No relationship found between... in the metadata. Please update the metadata"

R-Palazzo replied:
Yes, done in [76fcf86](https://github.com/sdv-dev/SDMetrics/pull/384/commits/76fcf862286df85acac9ee2977216ac5c72f5a30)

amontanez24 reviewed (approved):
:shipit:
